### ISSUE TITLE
Correct behaviour for apps using react-compiler

### ISSUE THREAD

**Issue by stevoland** [react-hook-form/react-hook-form#12298]
**Is your feature request related to a problem? Please describe.**

Experimenting with react-compiler, the only failing tests in our suite seem to be due to using parts of react-hook-form which break the "rules of react" ([eg: hook return values are immutable](https://react.dev/reference/rules/components-and-hooks-must-be-pure#return-values-and-arguments-to-hooks-are-immutable)) (ie: not limited to `watch`)

I've added react-compiler to the react-hook-form unit test code [here](https://github.com/stevoland/react-hook-form/compare/master...react-compiler) and see

```
Test Suites: 15 failed, 81 passed, 96 total
Tests:       35 failed, 836 passed, 871 total
```

**Describe the solution you'd like**

The library should be consistent and performant whether the user is compiling or not.

I think this can be achieved by adhering to the rules. eg: [this change](https://github.com/stevoland/react-hook-form/commit/1379b26490fe29dcf69869b8616cdf8d9112f36b) fixes `src/__tests__/useFormContext.test.tsx` (but there may be consequences or cases I'm not aware of.)

Happy to contribute but will need guidance from maintainers.

Thanks!

**Comment by bluebill1049** [comment]
Thanks for the feature request. Could you start to open that PR?

**Comment by stevoland** [comment]
Sure, I'll clean up a PR adding a new task `test:react-compiler` which doesn't run on ci.

I've dug in a bit more and am less optimistic about solving this in the short term. I suggest:

- A note in the docs to say that components that call `useForm` should opt-out of compilation with `'use no memo'` pragma
- Publish an eslint rule (or perhaps a babel plugin) to enforce the above

**Comment by bluebill1049** [comment]
Related #12310 In terms of `watch` API, perhaps there is something to change in terms of syntax but it would be considered a breaking change

```tsx
const { formState: { values } } = useForm({
  watch: 'inputName'
  // watch: ['inputName'] 
})
```

**Comment by stevoland** [comment]
Have [posted a rough babel plugin](https://gist.github.com/stevoland/a6cb19ed10c33e420b08b8e5c69b2336) that opts out functions that reference `useForm` from compilation. It seems sufficient to unblock my project and the rhf unit tests also pass with it.

**Comment by bluebill1049** [comment]
Thanks @stevoland. I think at the moment, most of the issue is around the `watch` API, perhaps leave a note in the doc to highlight with react complier to start with.

**Comment by rodrigofariow** [comment]
FYI
I can report having issues with other properties besides `watch`.
For me it was `control`. Can happen when the compiler memoized the usage of a `Controller` component for instance.
Side effect was that it would cause `field.onChange` to not update the value.

**Comment by jaydiablo** [comment]
As others have mentioned, we've also run into issues with react-hook-form and the react compiler (latest beta release as of this writing). I think a good first step for this library would be to install the React compiler eslint plugin and recommended ruleset (https://react.dev/learn/react-compiler#installing-eslint-plugin-react-compiler) and making sure all of those lints are corrected. I forked react-hook-form and tested this locally and these are the current complaints that the eslint plugin raises:

```
/react-hook-form/src/useForm.ts
   76:8   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
   77:5   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
   84:3   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
   86:16  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
   87:14  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  106:5   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  106:6   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  118:6   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  118:7   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  128:6   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  128:21  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  134:6   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  134:21  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  155:6   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  155:31  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  157:3   error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  157:65  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
  159:10  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler

/react-hook-form/src/useSubscribe.ts
  13:3  error  Ref values (the `current` property) may not be accessed during render. (https://react.dev/reference/react/useRef)  react-compiler/react-compiler
```

This is just what it found in the src files (ignoring other issues it found in the test files).

In our case, the `handleSubmit` function wasn't receiving the updated form values when the form was submitted.

**Comment by jaydiablo** [comment]
I forked the existing `react-compiler` branch and updated the packages to add the eslint plugin, then updated the eslint config, if you want to pull that into the existing react-compiler branch: https://github.com/jaydiablo/react-hook-form/commit/dd47bb542aebd78ecbf6e56d3cb0188012143d95

**Comment by jaydiablo** [comment]
Sorry for the rapid updates, re-reading the react-compiler docs and realized that the errors that the eslint plugin catches would be seen by the compiler as well, and it would skip over compilation of those sections, so correcting those errors I listed above probably wouldn't change anything in terms of the broken code that users are seeing.

In our case, we added the `"use no memo;"` directive to all components that have a `useForm` hook (basically what the babel plugin posted above does), which so far has solved issues we're seeing with react-hook-form and the compiler (in a small project however).

**Comment by badsyntax** [comment]
In our case we resolved a lot of compiler related issues by using `useWatch` and `useFormState` but unfortunately it seems formState is not updating correctly with errors. I've tried both `register()` and `Controller` approaches.

**Comment by ybelakov** [comment]
Any update on this?

**Comment by manuelpoelzl** [comment]
@badsyntax sounds like the same thing I'm currently facing:

This is the behaviour with react compiler, form errors won't be shown until the whole form is refreshed.
https://github.com/user-attachments/assets/1ccfbace-1d38-43f1-9c5e-a96cef005aa1

But without the react compiler everything works as expected and the form errors are shown immediately
https://github.com/user-attachments/assets/9d8301da-7a60-4181-8c50-cae9f2891567

**Comment by yslpn** [comment]
As an alternative solution, you can switch to TanStack Form, they promise full React Compiler support

https://tanstack.com/form/latest/docs/comparison

<img width="908" alt="Image" src="https://github.com/user-attachments/assets/5ec9ed66-30e3-445d-8c52-b5e499fae3d3" />

**Comment by brantphoto** [comment]
Hi there. Sorry, not trying to spam a thread, just planning a bit for fairly large codebase that has been transitioning to react-hook-form over the last year.

Soon we'll be to React 19 and also soon we'll be on the compiler. 

For context, just know that I know this is open source, and I appreciate the library, so take this with grain of salt: I'd like a little more information about the current status of compiler compatibility.

There is this PR, which appears promising, but also hasn't been rebased in 5 months (from what I can tell). That makes me wonder if it has been abandoned, if there is a blocker, etc (sorry I'm a worrier).
https://github.com/react-hook-form/react-hook-form/pull/12333

If possible, can I ask for a few things?:

- confirmation that this is the thread to watch to know when compatibility arrives.
- more insight into next steps in the draft PR
- potential tasks that could be done by others (if that iz viable)

If not, no worries, but in case any of that information comes quickly I'd appreciate it (notice I didn't ask for a timeline). Thanks.

**Comment by yslpn** [comment]
Hey everyone 👋,
now that [React Compiler RC](https://react.dev/blog/2025/04/21/react-compiler-rc) is officially out, just wondering if there are any updates or new thoughts regarding RHF compatibility?
Is there anything the community can help with to move things forward? Would love to hear any plans, blockers, or guidance for contributing! 🚀
Thanks again to everyone who's been working on this — really appreciate it!

**Comment by bluebill1049** [comment]
still wip will get to the latest RC version https://github.com/react-hook-form/react-hook-form/pull/12333

**Comment by ybelakov** [comment]
Any updates?

**Comment by RohovDmytro** [comment]
@bluebill1049 is there any workarounds or patches that can be applied meanwhile?

I am wrapping my head how to both enable react-compiler and use react-hook-form. Is it possible?

**Comment by trangcongthanh** [comment]
> I am wrapping my head how to both enable react-compiler and use react-hook-form. Is it possible?

You can use 'use no memo' to bypass complier for react-hook-form.

Example:
```
function Component() {
  'use no memo'
  const form = useForm()

  // rest of component
}
```

Someone above already made a babel plugin to add 'use no memo' for all instance of useForm in your project. You can check on it.

**Comment by theoludwig** [comment]
What exactly doesn't work with the React Compiler?
It seems like for simple usage of `useForm` hook, it works pretty good. At least in my testing, it works as it should with the React Compiler, and from what I see in the issue OP, most automated tests of `react-hook-form` work with the React compiler.

**Comment by Svish** [comment]
> What exactly doesn't work with the React Compiler?
> It seems like for simple usage of `useForm` hook, it works pretty good. At least in my testing, it works as it should witn the React Compiler, and from what I see in the issue OP, most automated tests of `react-hook-form` work with the React compiler.

Several things returned from `useForm` (and `useFormContext`) which are normally supposed to cause rerenders (like `watch`) gets memoized by the compiler which breaks all usage of those.

**Comment by lauri865** [comment]
> > What exactly doesn't work with the React Compiler?
> > It seems like for simple usage of `useForm` hook, it works pretty good. At least in my testing, it works as it should witn the React Compiler, and from what I see in the issue OP, most automated tests of `react-hook-form` work with the React compiler.
> 
> Several things returned from `useForm` (and `useFormContext`) which are normally supposed to cause rerenders (like `watch`) gets memoized by the compiler which breaks all usage of those. 

Just rename watch to useWatch and react compiler will handle it correctly. We use it (extensively) in production, no 'no use memo'.

**Comment by mwskwong** [comment]
> > > What exactly doesn't work with the React Compiler?
> > > It seems like for simple usage of `useForm` hook, it works pretty good. At least in my testing, it works as it should witn the React Compiler, and from what I see in the issue OP, most automated tests of `react-hook-form` work with the React compiler.
> > 
> > 
> > Several things returned from `useForm` (and `useFormContext`) which are normally supposed to cause rerenders (like `watch`) gets memoized by the compiler which breaks all usage of those.
> 
> Just rename watch to useWatch and react compiler will handle it correctly. We use it (extensively) in production, no 'no use memo'.

`watch()` is just one of several examples. There are other things like `formState` that works in similar fashion of `watch()`. To find them all is basically an adventure of trial and error, which is definitely not a good idea.

IMO, the safest way is to simply omit React Hook Form from React Compiler entirely at the current stage.

**Comment by RohovDmytro** [comment]
>  the safest way is to simply omit React Hook Form from React Compiler entirely at the current stage.

DO you also suggest to utilize `use no memo` or is there an alternative way?

**Comment by rodrigofariow** [comment]
> I am wrapping my head how to both enable react-compiler and use react-hook-form. Is it possible?

Hi everyone!
So, the way I'm allowing react compiler to still keeping "working", though in a sub optimal way, is to set `enableAssumeHooksFollowRulesOfReact=false` on the compiler config.

This will basically tell the compiler to **not memoize any component that contains any custom hooks.**
In short, if your component has `useEffect, useState, useMemo, etc`, you are okay. If you add something other than those, react compiler skips it.

This is not great but we still get some benefits in a bunch of components and allow us to keep using react-hook-form  (and any other problematic libraries) "without" worries.

Here is my config:

![Image](https://github.com/user-attachments/assets/480368ad-b2b6-4b55-bed2-6622032bdc17)

**Comment by ybelakov** [comment]
> > I am wrapping my head how to both enable react-compiler and use react-hook-form. Is it possible?
> 
> You can use 'use no memo' to bypass complier for react-hook-form.
> 
> Example:
> 
> ```
> function Component() {
>   'use no memo'
>   const form = useForm()
> 
>   // rest of component
> }
> ```
> 
> Someone above already made a babel plugin to add 'use no memo' for all instance of useForm in your project. You can check on it.

I tried 
```
 function useFormNoMemo(args) {
   'use no memo'
   const form = useForm(args)
 
   return form
 }
```

But this didn't work. I still had to put `'use no memo'` at the start of every components to make it work

**Comment by longzheng** [comment]
I created an ESLint plugin to help auto-add `'use no memo'` to components that use `useForm()` so I don't accidentally forget. https://www.npmjs.com/package/eslint-plugin-use-no-memo

**Comment by mauricedoepke** [comment]
I noticed renaming the watch function to useWatch like this, seems to make it work again:

```typescript
const { watch: useWatch } = useForm<>();

const thisWayWatchWorks = useWatch("myField")
```

**Comment by bluebill1049** [comment]
> I noticed renaming the watch function to useWatch like this, seems to make it work again:
> 
> const { watch: useWatch } = useForm<>();
> 
> const thisWayWatchWorks = useWatch("myField")

https://github.com/react-hook-form/react-hook-form/releases/tag/v8.0.0-alpha.5

**Comment by michaelhays** [comment]
> https://github.com/react-hook-form/react-hook-form/releases/tag/v8.0.0-alpha.5

Does this mean React Compiler support is coming soon? 😇

**Comment by Lamarcke** [comment]
> > https://github.com/react-hook-form/react-hook-form/releases/tag/v8.0.0-alpha.5
> 
> Does this mean React Compiler support is coming soon? 😇

I'm using the v8.0.0-alpha.5 version in production and didn't notice any issues. But i generally don't use controllers, only watch, register, etc and form context.

**Comment by Tehnix** [comment]
In the case of `watch` it seems the new convention is to switch to an imported `useWatch` (docs here https://react-hook-form.com/docs/usewatch) which keeps the watch function stable across renders, which renaming the returned `watch` from `useForm` to `useWatch` will not do (and thus still violates the rules of hooks).

Doesn't solve the other issues, but unblocked me and doesn't require any hacks.

So instead of:

```tsx
import { useForm } from 'react-hook-form';

// ... in your component
const {
  control,
  handleSubmit,
  watch: useWatch,
  formState: { errors, isValid, dirtyFields },
} = useForm<NicknameInputSchemaType>({
  resolver: zodResolver(nicknameSchema),
  mode: 'onChange',
});

const nicknameValue = useWatch('nickname', '');
```

You do:

```tsx
import { useForm, useWatch } from 'react-hook-form';

// ... in your component
const {
  control,
  handleSubmit,
  // NOTE: Don't get watch from the return values here.
  formState: { errors, isValid, dirtyFields },
} = useForm<NicknameInputSchemaType>({
  resolver: zodResolver(nicknameSchema),
  mode: 'onChange',
});

const nicknameValue = useWatch({ control, name: 'nickname', defaultValue: '' });
```

**Comment by timkindberg** [comment]
I wanted to get a fuller picture of what RHF apis were broken, and of those, which had valid alternatives (e.g. watch -> useWatch). Made this repo that tests all the apis and wrote up a report of the results.

https://github.com/timkindberg/rhf-compiler-compat

TLDR; Here are the key take-aways for devs wondering what they need to do for now until this is fixed in the library:

## Workarounds

For each breaking API, here are the recommended workarounds:

| Breaking API | Recommended Workaround | Notes |
|--------------|------------------------|-------|
| `form.watch('field')` | Replace with `useWatch({ name: 'field', control })` | Hook-based subscription, triggers re-renders properly |
| `form.watch()` (no args) | Replace with `useWatch({ control })` | Watches all fields via hook subscription |
| `useFormContext()` + watch | **No alternative** -- Use `'use no memo'` | Context consumers are memoized by compiler |
| `<Controller>` | **No alternative** -- Use `'use no memo'` | Control object interior mutability |
| `useController` | **No alternative** -- Use `'use no memo'` | Control object interior mutability |
| `getValues()` in render | Replace with `useWatch({ name: 'field', control })` for specific fields, OR move `getValues()` to callbacks/effects | Avoid calling during render; use hooks for reactive values |
| `reset()` | **No alternative** -- Use `'use no memo'` | Reset behavior breaks under memoization |
| `reset()` with new defaults | **No alternative** -- Use `'use no memo'` | Same as reset() |
| `watch` in useEffect deps | Replace watched value with `useWatch({ name: 'field', control })` | Effect deps need reactive values from hooks |
| Conditional fields via watch | Replace `form.watch('type')` with `useWatch({ name: 'type', control })` | Conditional rendering needs reactive hook values |
| Nested watch paths | Replace `form.watch('user.address.city')` with `useWatch({ name: 'user.address.city', control })` | Deep paths work with useWatch |
| `useFieldArray` + `watch` | Replace `form.watch('items')` with `useWatch({ name: 'items', control })` | Array watching needs hook subscription |
| `formState.isDirty` via context | Child should use `useFormState({ control })` with control passed as prop | Hook-based state access instead of context |

**Comment by bjornwiberg** [comment]
Adding to @rodrigofariow's 2024-10-21 report — confirming the Controller render-prop `field.onChange` no-op symptom in a Next.js 16 + `reactCompiler: true` app, and that it's a behaviour change between 7.73 and 7.75 (only visible in the production build).

### What changed (minimal example)

```tsx
function Form() {
  const form = useForm({ defaultValues: { picked: 'a' } });

  const onSubmit = async (data) => {
    await save(data);
    form.reset();           // ← key step
  };

  return (
    <form onSubmit={form.handleSubmit(onSubmit)}>
      <Controller
        control={form.control}
        name="picked"
        render={({ field }) => (
          <Chips
            value={field.value}
            onChange={(v) => field.onChange(v)}   // ← stops working in 7.75 (prod)
            options={['a', 'b']}
          />
        )}
      />
      <button type="submit">save</button>
    </form>
  );
}
```

Reproduction sequence:

1. Mount form, submit successfully → `form.reset()` runs.
2. Component stays mounted (e.g. user navigates to the same route, or the form is rendered inside a layout that survives the route change).
3. User clicks a different chip.

| RHF | Build  | Result                                                              |
| --- | ------ | ------------------------------------------------------------------- |
| 7.73| dev    | ✅ chip click updates `field.value`, form re-renders                |
| 7.73| prod (React Compiler) | ✅ same                                              |
| 7.75| dev    | ✅ same                                                             |
| 7.75| prod (React Compiler) | ❌ chip click is silently dropped — `field.value` does not update, form does not re-render, downstream `useWatch({ name: 'picked' })` still returns the previous value |

Without the `form.reset()` call in `onSubmit`, the same chip click works fine on 7.75 in prod. The bug only surfaces when the reset has happened and the component is still mounted on subsequent renders.

We hit this on two independent forms in the same codebase — both calling `form.reset()` after submit — and a third form using the same Controller render-prop pattern but never calling `form.reset()` is **not** affected. Consistent with `_fields`/`_names` being wiped in `_reset` and the compiler-memoized handler losing its registration target on the next click, though I haven't pinpointed the exact diff hunk.

### Workaround that fixed it for us

Drop the Controller wrapper for the affected field and drive the value with `useWatch` + `form.setValue`:

```tsx
const picked = useWatch({ control: form.control, name: 'picked' });
// ...
<Chips
  value={picked}
  onChange={(v) => form.setValue('picked', v, { shouldDirty: true })}
  options={['a', 'b']}
/>
```

Two notes for anyone copying this:

- **`form.watch(name)` is not a drop-in replacement** for the Controller's re-render subscription. We tried it first — the `setValue` call updated `_formValues` correctly, but the component never re-rendered. `useWatch({ control, name })` does subscribe; `form.watch(name)` returns the right value at the moment of the call but doesn't trigger a re-render on later changes.
- The repro really does need the `form.reset()` + still-mounted-component combination. Without one or the other, no manifestation.

### Versions

- `react-hook-form`: 7.75.0 (worked on 7.73.1 in the same app/env)
- `next`: 16 with `reactCompiler: true`
- `react`: 19

**Comment by Svish** [comment]
> Drop the Controller wrapper for the affected field and drive the value with useWatch + form.setValue

Would it not be better to replace the use of `Controller` with `useController` then?

**Comment by bjornwiberg** [comment]
Good call — just tested it on the same prod-build repro and `useController` works:

```tsx
const { field } = useController({ control: form.control, name: 'picked' });

<Chips
  value={field.value}
  onChange={(v) => field.onChange(v)}
  options={['a', 'b']}
/>
```

Same `form.reset()` + still-mounted-component + chip-click sequence as before — passes on 7.75 + React Compiler prod build across all 5 device projects in our Playwright suite. So the bug isn't in `field.onChange` itself; it's specifically in the `<Controller render={...}>` render-prop wrapper. The bare hook is fine.

Updated take-away for anyone hitting this:

| Pattern                                                         | 7.75 + React Compiler prod |
| --------------------------------------------------------------- | -------------------------- |
| `<Controller render={({field}) => <X onChange={field.onChange}/>}>` | ❌                         |
| `const { field } = useController(...); <X onChange={field.onChange}/>` | ✅                    |
| `useWatch(...) + form.setValue(...)`                           | ✅                         |

`useController` is the cleaner of the two workarounds since it preserves validation rules / `field.ref` for anyone who needs them. The `useWatch + setValue` path is only really needed when you'd already be skipping `Controller` for a pure value+onChange field with no rules.

Thanks @Svish — going to swap our codebase over to `useController`.

**Comment by bluebill1049** [comment]
We’ll be making all the necessary changes to ensure compatibility with React Hook Form v8. You can try the latest beta version in the meantime.

Migration guide: https://react-hook-form.com/migrate-v7-to-v8

Please let us know if you run into any issues while testing the beta.

**Comment by timkindberg** [comment]
Quick follow-up now that v8 beta is out.

I re-ran my RHF + React Compiler compat harness on `react-hook-form@8.0.0-beta.2` and put the full write-up here:
https://github.com/timkindberg/rhf-compiler-compat  

Report: https://github.com/timkindberg/rhf-compiler-compat/blob/main/REPORT-rhf-8.0.0-beta.2.md

### Correctness: better on v8

Same compiler (`babel-plugin-react-compiler@1.0.0`), only RHF version changed.

- v7.75: ❌ `37/42` with compiler
- v8 beta: ✅ `42/42` with compiler

So the previously failing patterns in this thread are green in my harness on v8 beta.

### “compatible” != “optimized”

I think this is the important caveat: components using `useForm().watch()` still get compiler bailout (`Use of incompatible library`) in current compiler logic, so those components are correct but not memoized by the compiler.

That detail (plus emitted-output examples) is in the report.

### Why I think react-compiler's hard-coded bailout is no longer for v8

I forced memoization of `useForm().watch()` (via a local re-export trick to bypass source-string matching) and got:

- v7.75 + forced memoization: fails ❌
- v8 beta + forced memoization: passes ✅

So IMO v8 fixed the underlying safety issue, but compiler bailout logic is still conservative.

### Requests / questions

1. Maybe worth syncing with React Compiler team on whether the RHF `watch` bailout can be version-gated or removed for v8?
2. Maybe docs can call out that `form.watch()` may still be skipped by compiler optimization, and `useWatch({ control })` is the safer “optimized path” today.

Happy to help with doc PRs or more targeted repros if useful.

**Comment by sKopheK** [comment]
tried `react-hook-form@8.0.0-beta.3` and `getValues` from `useFormContext` is not stable there - actually causes infinite component rerenders just when referencing it

**Comment by bluebill1049** [comment]
> tried `react-hook-form@8.0.0-beta.3` and `getValues` from `useFormContext` is not stable there - actually causes infinite component rerenders just when referencing it

Share code usage.

**Comment by sKopheK** [comment]
> > tried `react-hook-form@8.0.0-beta.3` and `getValues` from `useFormContext` is not stable there - actually causes infinite component rerenders just when referencing it
> 
> Share code usage.

wasn't able to reproduce it anymore, so could have been some other property causing the behavior probably, even though using `@simbathesailor/use-what-changed` i definitely have seen getValues being changed every render 🤷 so upgrading when app was running and then turbopack didn't rebuild it properly is also probability. however, you can safely ignore what i wrote before, thanks for reply

**Comment by bluebill1049** [comment]
Thanks for the update! If you’re able to reproduce the issue, please let me know. I’ll investigate it further.

**Comment by xiongemi** [comment]
is there any update to merge in https://github.com/react-hook-form/react-hook-form/pull/12333
and publish the official v8?
